packages/terser/src/index.bzl

```diff
@@ -0,0 +1,126 @@
+        """,
+    ),
+    "sourcemap": attr.bool(
+        doc = "TODO",
+        default = True,
+    ),
+    "terser_bin": attr.label(
+        default = Label("@npm//@bazel/terser/bin:terser"),
+        executable = True,
+        cfg = "host",
+    ),
+    "deps": attr.label_list(),
+}
+
+TERSER_OUTS = {
+    "optimized": "%{name}.js",
+}
+
+# Converts a dict to a struct, recursing into a single level of nested dicts.
+# This allows users of compile_ts to modify or augment the returned dict before
+# converting it to an immutable struct.
+def _dict_to_struct(d):
+    for key, value in d.items():
+        if type(value) == type({}):
+            d[key] = struct(**value)
+    return struct(**d)
+
+def run_terser(ctx, srcs, output):
+    """TODO: doc"""
+
+    # CLI arguments; see https://www.npmjs.com/package/terser#command-line-usage
+    args = ctx.actions.args()
+    args.add_all([src.path for src in srcs])
+
+    outputs = [output]
+    args.add_all(["--output", output])
+
+    # TODO: also check the env.DEBUG variable
+    debug = ctx.attr.debug
+    if debug:
+        args.add("--debug")
+        args.add("--beautify")
+
+    # See https://github.com/terser-js/terser#minify-options
+    # TODO: give user appropriate control over options
+    minify_options = {
+        "compress": {
+            "global_defs": {"ngDevMode": False, "ngI18nClosureMode": False},
+            "keep_fnames": not debug,
+            "passes": 3,
+            "pure_getters": True,
+            "reduce_funcs": not debug,
+            "reduce_vars": not debug,
+            "sequences": not debug,
+        },
+        "mangle": not debug,
+    }
+
+    if ctx.attr.sourcemap:
+        map_output = ctx.actions.declare_file(output.basename + ".map", sibling = output)
+        outputs.append(map_output)
+
+        # Source mapping options are comma-packed into one argv
+        # see https://github.com/terser-js/terser#command-line-usage
+        source_map_opts = ["includeSources", "base=" + ctx.bin_dir.path]
+        #if in_source_map:
+        #    source_map_opts.append("content=" + in_source_map.path)
+        #    inputs.append(in_source_map)
+
+        # This option doesn't work in the config file, only on the CLI
+        args.add_all(["--source-map", ",".join(source_map_opts)])
+        minify_options["sourceMap"] = {"filename": map_output.path}
+
+    opts = ctx.actions.declare_file("_%s.minify_options.json" % ctx.label.name)
+    ctx.actions.write(opts, _dict_to_struct(minify_options).to_json())
+    args.add_all(["--config-file", opts.path])
+
+    ctx.actions.run(
+        inputs = srcs + [opts],
+        outputs = outputs,
+        executable = ctx.executable.terser_bin,
+        arguments = [args],
+        progress_message = "Optimizing JavaScript %s [terser]" % (output.short_path),
+    )
+    return outputs
+
+def _terser(ctx):
+    modules = collect_js_modules(ctx)
+    outputs = run_terser(ctx, modules.sources, ctx.outputs.optimized)
```

Comment:
What if terser is invoked without using `JSModuleInfo`? I thought we'd be running terser on each provider individually, something such as what was started at https://github.com/jbedard/rules_nodejs/commit/58c8d83813a5972ceb811d56745f669e0e2eb102#diff-86d64554a52a8e537d18c6619fcedc10R129

Reply:
That seems like a strange and unexpected shape as well since terser would then have multiple outputs and would need to opinionated in the output name. With a single output from either `JSModuleInfo` (if it exists) or from DefaultInfo files (fallback) then it would be just be named `{name}.js` by default. To use other providers you would put a JS provider filter rule before terser that would take a provider such as `JSNamedModuleInfo` and output it as `JSModuleInfo` for terser to consume. Something similar is already done in g3 for ts_devserver which does not know how to find the named js on its own.

Reply:
Yeah the output names would definitely be a problem, and I guess putting a single `{name}.js` file into both providers like this solves that.

Would it ever be possible for a user to expect `DefaultInfo` to be used instead of `JSModuleInfo`, and that causing unexpected output in `DefaultInfo`? I guess just adding another step (putting things into a `filegroup`?) would fix that though.

Reply:
Initial terser rule was landed on master and uses some bazel magic @alexeagle discovered to get named outputs for both the `{name}.js` and `{name}.js.map` if sourcemap is enabled. The landed version uses DefaultInfo only and I rebased on this branch for it to handle the JSModuleInfo provider being worked on here as well.

Default behavior is use JSModuleInfo if found otherwise fallback to DefaultInfo, however, we're leaning towards not using apsects to collect JSTransitiveModuleInfo so JSModuleInfo will always provide all direct & transitive deps. This means that if you really want to force the usage of DefaultInfo you can the `src` through a `filegroup` which will filter it out and because no aspect is run it won't be found.